kani-compiler/src/kani_middle/transform/loop_contracts.rs

```diff
@@ -0,0 +1,313 @@
+
+//! This module contains code related to the MIR-to-MIR pass to enable loop contracts.
+//!
+
+use crate::kani_middle::KaniAttributes;
+use crate::kani_middle::codegen_units::CodegenUnit;
+use crate::kani_middle::find_fn_def;
+use crate::kani_middle::transform::TransformationType;
+use crate::kani_middle::transform::body::{InsertPosition, MutableBody, SourceInstruction};
+use crate::kani_queries::QueryDb;
+use crate::stable_mir::CrateDef;
+use rustc_middle::ty::TyCtxt;
+use rustc_span::Symbol;
+use stable_mir::mir::mono::Instance;
+use stable_mir::mir::{
+    AggregateKind, BasicBlockIdx, Body, ConstOperand, Operand, Place, Rvalue, StatementKind,
+    Terminator, TerminatorKind, VarDebugInfoContents,
+};
+use stable_mir::ty::{FnDef, MirConst, RigidTy};
+use std::collections::{HashMap, HashSet, VecDeque};
+use std::fmt::Debug;
+
+use super::TransformPass;
+
+/// This pass will perform the following operations:
+/// 1. Replace the body of `kani_register_loop_contract` by `kani::internal::run_contract_fn`
+///    to invoke the closure.
+///
+/// 2. Transform loops with contracts from
+///    ```ignore
+///    bb_idx: {
+///         loop_head_stmts
+///         _v = kani_register_loop_contract(move args) -> [return: terminator_target];
+///    }
+///
+///    ...
+///    loop_body_blocks
+///    ...
+///
+///    loop_latch_block: {
+///         loop_latch_stmts
+///         goto -> bb_idx;
+///    }
+///    ```
+///    to blocks
+///    ```ignore
+///    bb_idx: {
+///         _v = true
+///         goto -> terminator_target
+///    }
+///
+///    ...
+///    loop_body_blocks
+///    ...
+///
+///    loop_latch_block: {
+///         loop_latch_stmts
+///         goto -> bb_new_loop_latch;
+///    }
+///
+///    bb_new_loop_latch: {
+///         loop_head_body
+///         _v = kani_register_loop_contract(move args) -> [return: terminator_target];
+///    }
+///    ```
+#[derive(Debug, Default)]
+pub struct LoopContractPass {
+    /// Cache KaniRunContract function used to implement contracts.
+    run_contract_fn: Option<FnDef>,
+    /// The map from original loop head to the new loop latch.
+    /// We use this map to redirect all original loop latches to a new single loop latch.
+    new_loop_latches: HashMap<usize, usize>,
+}
+
+impl TransformPass for LoopContractPass {
+    /// The type of transformation that this pass implements.
+    fn transformation_type() -> TransformationType
+    where
+        Self: Sized,
+    {
+        TransformationType::Stubbing
+    }
+
+    fn is_enabled(&self, query_db: &QueryDb) -> bool
+    where
+        Self: Sized,
+    {
+        query_db.args().unstable_features.contains(&"loop-contracts".to_string())
+    }
+
+    /// Run a transformation pass on the whole codegen unit.
+    fn transform(&mut self, tcx: TyCtxt, body: Body, instance: Instance) -> (bool, Body) {
+        match instance.ty().kind().rigid().unwrap() {
+            RigidTy::FnDef(_func, args) => {
```

Comment:
What about closures (and co-routines)... won't they be `ClosureDef`?

Reply:
Good points. I think the current implementation also support loop contracts in closures and co-routines. I will enable them and add tests for them. Tracked in #3599.